src/langchain_google_memorystore_redis/doc_saver.py

```diff
@@ -0,0 +1,113 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import uuid
+from typing import Optional, Sequence, Set, Union
+
+import redis
+from langchain_core.documents.base import Document
+
+
+class MemorystoreDocumentSaver:
+    """Document Saver for Cloud Memorystore for Redis database."""
+
+    def __init__(
+        self,
+        client: redis.Redis,
+        key_prefix: str,
+        content_field: str,
+        metadata_fields: Optional[Set[str]] = None,
+    ):
+        """Initializes the Document Saver for Memorystore for Redis.
+
+        Args:
+            client: A redis.Redis client object.
+            key_prefix: A prefix for the keys to store Documents in Redis.
+            content_field: The field of the hash that Redis uses to store the
+                page_content of the Document.
+            metadata_fields: The metadata fields of the Document that will be
+                stored in the Redis. If None, Redis stores all metadata fields.
+        """
+
+        self._redis = client
+        if not key_prefix:
+            raise ValueError("key_prefix must not be empty")
+        self._key_prefix = key_prefix
+        self._content_field = content_field
+        self._metadata_fields = metadata_fields
+
+    def add_documents(
+        self,
+        documents: Sequence[Document],
+        ids: Optional[Sequence[str]] = None,
+        batch_size: int = 1000,
+    ) -> None:
+        """Save a list of Documents to Redis.
+
+        Args:
+            documents: A List of Documents.
+            ids: The list of suffixes for keys that Redis uses to store the
+                Documents. If specified, the length of the IDs must be the same
+                as Documents. If not specified, random UUIDs appended after
+                prefix are used to store each Document.
+            batch_size: The number of documents to process in a single batch
+                operation. This parameter helps manage memory and performance
+                when adding a large number of documents. Defaults to 1000.
+        """
+        if ids and len(documents) != len(ids):
+            raise ValueError("The length of documents must match the length of the IDs")
+        if batch_size <= 0:
+            raise ValueError("batch_size must be greater than 0")
+
+        doc_ids = ids if ids else [str(uuid.uuid4()) for _ in documents]
+        doc_ids = [self._key_prefix + doc_id for doc_id in doc_ids]
+
+        pipeline = self._redis.pipeline(transaction=False)
+        for i, doc in enumerate(documents):
+            mapping = self._filter_metadata_by_fields(doc.metadata)
+            mapping.update({self._content_field: doc.page_content})
+
+            # Remove existing key in Redis to avoid reusing the doc ID.
```

Comment:
Shouldn't this error instead of deleting the old document?

Reply:
That sounds good. Updated